README.md
# Watchtower

Watchtower uses the [Elastic Stack](https://www.elastic.co/products/) and [Slack-Watchman](https://github.com/PaperMtn/slack-watchman) by [@PaperMtn](https://github.com/PaperMtn).

Slack is a great tool for easy and convenient team collaboration and growing rapidly in popularity with home office becoming more widespread. As with other collaboration platforms, i.e. GitHub, it may happen that sensitive information get posted by accident or convenience while it should have never been posted.

Based on an adjustable ruleset Slack-Watchman will run on a daily basis and detect if and when sensitive information was posted and provide a permalink to the corresponding post for analysis.

Logstash will ingest the JSON based logfiles and feed them into an Elasticsearch index while a pre-configured Kibana dashboard will beautifully visualize all the captured detections.

![Watchtower](dashboard.png)

## Prerequisites
Watchtower can be installed on top of a previously installed Debian 10.x and will automatically download all required packages, such as ...
```
curl
docker.io
docker-compose
grc
jq
pwgen
```
Where possible the installation will be as less intrusive as possible. Only `/etc/crontab` and `/etc/sysctl.conf` will be adjusted to automatically run Slack-Watchman and prepare for a dockerized Elastic Stack.

## Setup
The following steps must be done as `root` (`su -`).

Clone this repository into `/opt/watchtower`:
```
git clone https://github.com/telekom-security/watchtower /opt/watchtower
```

Run the `setup.sh` script:
```
cd /opt/watchtower
./setup.sh
```

Please be patient during the setup procedure, all required docker images will be downloaded and started to generate all the necessary folders, passwords and certificates for the Elastic Stack. Once the configuration is completed the Elastic Stack will be started and the pre-configured Kibana objects imported so you can start right away.

Please store all the passwords in a safe place and delete the temporary password file from `/data/elastic/conf/passwords`.

## Configuration
You need to register Slack-Watchman as an app for your Slack workspace. Follow this [Requirements Guide](https://github.com/PaperMtn/slack-watchman#requirements) to successfully register your app in Slack and retrieve the required API token.

Once you retrieved the API token you simply add it the `sw-runner.sh`
```
# fuSWRUN "Workspace_Name" "OAuthToken"
fuSWRUN "slack_workspace" "xoxp-1111111111111-2222222222222-3333333333333-444be444444444444444444444444444"
```
The `Workspace_Name` is a name of your choosing and will determine the logfile name (*.json suffix will be added automatically).
The `OAuthToken` is the API token you will be provided via the Slack app settings.
Following this method you can add as many workspaces as you like and all will be monitored on a daily basis.

## Run Slack-Watchman
If this is a primary run with Slack-Watchman you can set the timeframe in `sw-runner.sh` to `--timeframe a` which is the default and will scan all past events. It's recommended however to change this to `--timeframe d` which will only scan the previous 24 hours.

For testing purposes you can run Slack-Watchman simply via `sw-runner.sh`. But keep in mind that each run will account against the API, once too hot your app will be put into cool off mode. Running once daily per registered Slack app is recommended.

## Upgrades
To upgrade just follow these simple steps from within the repo folder:
```
git pull
docker-compose -f docker/build.yml pull
./stop.sh
./start.sh
```

## Licenses
The software Watchtower is built on uses the following licenses.
<br>GPLv3: [Github-Watchman](https://github.com/PaperMtn/github-watchman/blob/master/LICENSE)
<br>GPLv3: [Gitlab-Watchman](https://github.com/PaperMtn/gitlab-watchman/blob/master/LICENSE)
<br>GPLv3: [Slack-Watchman](https://github.com/PaperMtn/slack-watchman/blob/master/LICENSE)
<br>Apache 2 License: [elasticsearch](https://github.com/elasticsearch/elasticsearch/blob/master/LICENSE.txt), [logstash](https://github.com/elasticsearch/logstash/blob/master/LICENSE), [kibana](https://github.com/elasticsearch/kibana/blob/master/LICENSE.md), [docker](https://github.com/docker/docker/blob/master/LICENSE)
<br> Other: [Debian licensing](https://www.debian.org/legal/licenses/)



CHANGELOG.md
# Changelog

# 20210111
## Add Slack-Watchman 3.0.10
Slack-Watchman 3.0.10 uses different field names which will require you to download `watchtower_kibana_objects.ndjson.zip` and import it into Kibana.
All Kibana objects were adjusted accordingly.

## Update Elastic Stack to 7.10.1
There might be breaking changes, if in doubt do not upgrade (`git pull`).

## Integration of Github and Gitlab Watchman
I just started to integrate Github- and Gitlab Watchman into the Docker image, no ELK or runnter integration yet (todo).



setup.sh
#!/bin/bash

# Vars
myPACKAGES="curl docker.io docker-compose git grc jq pwgen"
myINSTPATH="/opt/watchtower"
myGITREPO="https://github.com/telekom-security/watchtower"
myCRONJOBS="
# Run daily
0 6 * * *      root    /opt/watchtower/runner.sh
"

# Installer
echo "### Watchtower Installer"
echo

# Got root?
myWHOAMI=$(whoami)
if [ "$myWHOAMI" != "root" ]
  then
    echo
    echo "### Need to run as root ..."
    echo
    exit
fi

# Do not run again if certificate bundle exists already
if [ -f /data/elastic/certs/bundle.zip ];
  then
    echo
    echo "### Certificate bundle already exists. Aborting."
    echo
    exit
fi

# Elasticsearch needs adjustments to kernel settings
myVM_MAX_MAP_COUNT=$(grep "vm.max_map_count" /etc/sysctl.conf)
if [ "$myVM_MAX_MAP_COUNT" == "" ];
  then
    echo
    echo "### Now patching sysctl.conf ..."
    echo
    echo "vm.max_map_count = 262144" | tee -a /etc/sysctl.conf 
    echo
    echo "### Reloading Kernel Parameter ..."
    echo
    sysctl -p
  else
    echo "$myVM_MAX_MAP_COUNT is already set."
    echo
fi

# Check for deps
myINST=""
for myDEPS in $myPACKAGES;
do
  myOK=$(dpkg -s $myDEPS | head -n2 | grep ok | awk '{ print $3 }');
  if [ "$myOK" != "ok" ]
    then
      myINST=$(echo $myINST $myDEPS)
  fi
done
if [ "$myINST" != "" ]
  then
    echo
    echo "### Need to install some missing packages ..."
    echo
    apt-get update -y
    for myDEPS in $myINST;
    do
      apt-get install $myDEPS -y
    done
  else
    echo
    echo "### All dependencies are met."
    echo
fi

# Check if cloned to /opt/watchtower
myPATH=$(pwd)
if [ "$myPATH" != "$myINSTPATH" ];
  then
    echo
    echo "### Watchtower needs to be installed into $myINSTPATH."
    echo "### Cloning and restarting setup from correct path."
    echo
    mkdir -p /opt
    cd /opt
    git clone $myGITREPO
    cd $myINSTPATH
    ./$0
    exit
fi

# Create folders
echo 
echo "### Creating folders ..."
echo
mkdir -vp /data/elastic/{certs,conf,log,data} \
          /data/runner
chmod -R 770 /data
chown -R 1000:0 /data

# Pull images
echo
echo "### Pulling docker images, please be patient."
echo
docker-compose -f docker/build.yml pull

# Start elasticsearch for the first time to gen certs and passwords
echo
echo "### Running Elasticsearch for the first time, please be patient while generating certificates and passwords."
echo
docker-compose -f docker/elasticsearch/docker-compose.yml up -d
docker logs -f elasticsearch 2>&1 | grep -m 1 "\[GREEN\]"
docker-compose -f docker/elasticsearch/docker-compose.yml down -v

# Convert passwords, so we can source them as vars
cat /data/elastic/conf/passwords | grep "PASSWORD" | cut -d " " -f 2- | tr -d " " > /data/elastic/conf/passwords.source
source /data/elastic/conf/passwords.source

# Replace passwords in docker-compose.yml, so Elastic Stack can start properly just out of the box
sed -i '/ELASTICSEARCH_USERNAME: kibana_system/!b;n;c\      ELASTICSEARCH_PASSWORD: '$kibana_system'' docker/docker-compose.yml
sed -i '/ELASTICSEARCH_USERNAME: elastic/!b;n;c\      ELASTICSEARCH_PASSWORD: '$elastic'' docker/docker-compose.yml
sed -i "s/ELASTICSEARCH_OBJECTS_ENCRYPTION_KEY.*$/ELASTICSEARCH_OBJECTS_ENCRYPTION_KEY: $(pwgen -cnsB 32 1)/" docker/docker-compose.yml
sed -i "s/ELASTICSEARCH_SECURITY_ENCRYPTION_KEY.*$/ELASTICSEARCH_SECURITY_ENCRYPTION_KEY: $(pwgen -cnsB 32 1)/" docker/docker-compose.yml
sed -i "s/ELASTICSEARCH_REPORTING_ENCRYPTION_KEY.*$/ELASTICSEARCH_REPORTING_ENCRYPTION_KEY: $(pwgen -cnsB 32 1)/" docker/docker-compose.yml

# Remove passwords in source format
rm /data/elastic/conf/passwords.source

# Add cronjob
myCHECK=$(grep "watchtower" /etc/crontab)
if [ "$myCHECK" == "" ];
  then
    echo
    echo "### Now adding cronjob ..."
    echo "$myCRONJOBS" | tee -a /etc/crontab
    echo
  else
    echo "### Cronjob is already set."
    echo
fi

# Ready to start and import objects
./start.sh
echo
echo "### Waiting for Kibana to be healthy."
echo -n "### Please be patient "
while true;
  do
    myCHECK=$(docker ps | grep kibana | grep healthy | wc -l)
    if [ "$myCHECK" == "1" ];
      then
	echo
        echo "### Kibana is alive."
	echo
        ./import_kibana-objects.sh elastic $elastic kibana-objects.tgz
        break
      else
        echo -n "."
        sleep 2
    fi
done

# Done
echo "### Kibana superuser: elastic / password: $elastic"
echo "### All generated passwords are stored in /data/elastic/conf/passwords."
echo "### Retrieve the passwords, store them in a safe place and delete the passwords file."
echo "### Done."
echo


runner.sh
#!/bin/bash

# Got root?
myWHOAMI=$(whoami)
if [ "$myWHOAMI" != "root" ]
  then
    echo "### Need to run as root ..."
    echo
    exit
fi

### Define Slack-Watchman runner
fuSWRUN () {
local SLACK_WATCHMAN_NAME="$1"
local SLACK_WATCHMAN_TOKEN="$2"

# Changing the prefix will break logstash log ingestion
local SLACK_WATCHMAN_LOGFILE="slack_wm_$(echo $SLACK_WATCHMAN_NAME | tr -d " ").json"

# Options
## Timeframe: --timeframe {d,w,m,a}
## Query: [--users] [--channels] [--pii] [--financial] [--tokens] [--files] [--custom CUSTOM]
local SLACK_WATCHMAN_COMMAND="slack-watchman --timeframe a --all --output stdout | tee $SLACK_WATCHMAN_LOGFILE"

echo "### Now processing: $SLACK_WATCHMAN_NAME"
docker run -t --name runner \
              --rm \
              --read-only \
              -v /data/runner/log:/var/log/runner:rw \
              --env SLACK_WATCHMAN_TOKEN="$SLACK_WATCHMAN_TOKEN" \
              dtagdevsec/runner-wt:latest ash -c "$SLACK_WATCHMAN_COMMAND"

echo "### Done processing: $SLACK_WATCHMAN_NAME"
echo
}

# fuSWRUN "Workspace_Name" "OAuthToken"
fuSWRUN "Workspace_Name" "xoxp-1111111111111-2222222222222-3333333333333-44444444444444444444444444444444"


factory_reset.sh
#!/bin/bash

# Got root?
myWHOAMI=$(whoami)
if [ "$myWHOAMI" != "root" ]
  then
    echo "### Need to run as root ..."
    exit
fi

# Clear all settings, files, passwords, certs for Slack-Watchtower
echo "### Now removing all contents of /data."
echo "### This cannot be undone."
rm -rI /data/elastic /data/slack-watchman

# Replace passwords in docker-compose.yml with CHANGEME
sed -i '/ELASTICSEARCH_USERNAME: kibana_system/!b;n;c\      ELASTICSEARCH_PASSWORD: CHANGEME' docker/docker-compose.yml
sed -i '/ELASTICSEARCH_USERNAME: elastic/!b;n;c\      ELASTICSEARCH_PASSWORD: CHANGEME' docker/docker-compose.yml
sed -i "s/ELASTICSEARCH_OBJECTS_ENCRYPTION_KEY.*$/ELASTICSEARCH_OBJECTS_ENCRYPTION_KEY: CHANGEME/" docker/docker-compose.yml
sed -i "s/ELASTICSEARCH_SECURITY_ENCRYPTION_KEY.*$/ELASTICSEARCH_SECURITY_ENCRYPTION_KEY: CHANGEME/" docker/docker-compose.yml
sed -i "s/ELASTICSEARCH_REPORTING_ENCRYPTION_KEY.*$/ELASTICSEARCH_REPORTING_ENCRYPTION_KEY: CHANGEME/" docker/docker-compose.yml


export_kibana-objects.sh
#!/bin/bash
# Export all Kibana objects through Kibana Saved Objects API
# Make sure ES is available
myES="https://127.0.0.1:9200/"
myKIBANA="https://127.0.0.1:5601/"
myCURL="curl -u "$1":"$2" --insecure -s"

# Check if Kibana API is available
echo "### Waiting for Kibana API to be reachable."
echo -n "### Please be patient "
while true;
  do
    myCHECK=$($myCURL $myKIBANA/api | grep 404 | wc -l)
    if [ "$myCHECK" == 1 ];
      then
        echo
        echo "### Kibana API is available, now exporting objects."
        break
      else
        sleep 2
        echo -n "."
    fi
done

# Set vars
myDATE=$(date +%Y%m%d%H%M)
myINDEXCOUNT=$($myCURL -XGET ''$myKIBANA'api/saved_objects/_find?type=index-pattern' | jq '.saved_objects[].attributes' | tr '\\' '\n' | grep "scripted" | wc -w)
myINDEXID=$($myCURL -XGET ''$myKIBANA'api/saved_objects/_find?type=index-pattern' | jq '.saved_objects[].id' | tr -d '"')
myDASHBOARDS=$($myCURL -XGET ''$myKIBANA'api/saved_objects/_find?type=dashboard&per_page=500' | jq '.saved_objects[].id' | tr -d '"')
myVISUALIZATIONS=$($myCURL -XGET ''$myKIBANA'api/saved_objects/_find?type=visualization&per_page=500' | jq '.saved_objects[].id' | tr -d '"')
mySEARCHES=$($myCURL -XGET ''$myKIBANA'api/saved_objects/_find?type=search&per_page=500' | jq '.saved_objects[].id' | tr -d '"')
myCONFIGS=$($myCURL -XGET ''$myKIBANA'api/saved_objects/_find?type=config&per_page=500' | jq '.saved_objects[].id' | tr -d '"')
myCOL1="[0;34m"
myCOL0="[0;0m"

# Let's ensure normal operation on exit or if interrupted ...
function fuCLEANUP {
  rm -rf patterns/ dashboards/ visualizations/ searches/ configs/
}
trap fuCLEANUP EXIT

# Export index patterns
mkdir -p patterns
echo $myCOL1"### Now exporting"$myCOL0 $myINDEXCOUNT $myCOL1"index pattern fields." $myCOL0
$myCURL -XGET ''$myKIBANA'api/saved_objects/index-pattern/'$myINDEXID'' | jq '. | {attributes, references}' > patterns/$myINDEXID.json &
echo

# Export dashboards
mkdir -p dashboards
echo $myCOL1"### Now exporting"$myCOL0 $(echo $myDASHBOARDS | wc -w) $myCOL1"dashboards." $myCOL0
for i in $myDASHBOARDS;
  do
    echo $myCOL1"###### "$i $myCOL0
    $myCURL -XGET ''$myKIBANA'api/saved_objects/dashboard/'$i'' | jq '. | {attributes, references}' > dashboards/$i.json &
  done;
echo

# Export visualizations
mkdir -p visualizations
echo $myCOL1"### Now exporting"$myCOL0 $(echo $myVISUALIZATIONS | wc -w) $myCOL1"visualizations." $myCOL0
for i in $myVISUALIZATIONS;
  do
    echo $myCOL1"###### "$i $myCOL0
    $myCURL -XGET ''$myKIBANA'api/saved_objects/visualization/'$i'' | jq '. | {attributes, references}' > visualizations/$i.json &
  done;
echo

# Export searches
mkdir -p searches
echo $myCOL1"### Now exporting"$myCOL0 $(echo $mySEARCHES | wc -w) $myCOL1"searches." $myCOL0
for i in $mySEARCHES;
  do
    echo $myCOL1"###### "$i $myCOL0
    $myCURL -XGET ''$myKIBANA'api/saved_objects/search/'$i'' | jq '. | {attributes, references}' > searches/$i.json &
  done;
echo

# Export configs
mkdir -p configs
echo $myCOL1"### Now exporting"$myCOL0 $(echo $myCONFIGS | wc -w) $myCOL1"configs." $myCOL0
for i in $myCONFIGS;
  do
    echo $myCOL1"###### "$i $myCOL0
    $myCURL -XGET ''$myKIBANA'api/saved_objects/config/'$i'' | jq '. | {attributes, references}' > configs/$i.json &
  done;
echo

# Wait for background exports to finish
wait

# Building tar archive
echo $myCOL1"### Now building archive"$myCOL0 "kibana-objects_"$myDATE".tgz"
tar cvfz kibana-objects_$myDATE.tgz patterns dashboards visualizations searches configs > /dev/null

# Stats
echo
echo $myCOL1"### Statistics"
echo $myCOL1"###### Exported"$myCOL0 $myINDEXCOUNT $myCOL1"index patterns." $myCOL0
echo $myCOL1"###### Exported"$myCOL0 $(echo $myDASHBOARDS | wc -w) $myCOL1"dashboards." $myCOL0
echo $myCOL1"###### Exported"$myCOL0 $(echo $myVISUALIZATIONS | wc -w) $myCOL1"visualizations." $myCOL0
echo $myCOL1"###### Exported"$myCOL0 $(echo $mySEARCHES | wc -w) $myCOL1"searches." $myCOL0
echo $myCOL1"###### Exported"$myCOL0 $(echo $myCONFIGS | wc -w) $myCOL1"configs." $myCOL0
echo


import_kibana-objects.sh
#!/bin/bash
# Import Kibana objects
# Make sure ES is available
myES="https://127.0.0.1:9200/"
myKIBANA="https://127.0.0.1:5601/"
myCURL="curl -u "$1":"$2" --insecure -s"

# Check if Kibana API is available
echo "### Waiting for Kibana API to be reachable."
echo -n "### Please be patient "
while true;
  do
    myCHECK=$($myCURL $myKIBANA/api | grep 404 | wc -l)
    if [ "$myCHECK" == 1 ];
      then
	echo
        echo "### Kibana API is available, now importing objects."
	sleep 5
        break
      else
        sleep 2
        echo -n "."
    fi
done

# Set vars
myDUMP=$3
myCOL1="[0;34m"
myCOL0="[0;0m"

# Let's ensure normal operation on exit or if interrupted ...
function fuCLEANUP {
  rm -rf patterns/ dashboards/ visualizations/ searches/ configs/
}
trap fuCLEANUP EXIT

# Check if parameter is given and file exists
if [ "$myDUMP" = "" ];
  then
    echo $myCOL1"### Please provide a backup file name."$myCOL0 
    echo $myCOL1"### import_kibana-objects.sh <kibana-objects.tgz>"$myCOL0
    echo 
    exit
fi
if ! [ -a $myDUMP ];
  then
    echo $myCOL1"### File not found."$myCOL0 
    exit
fi

# Unpack tar
tar xvfz $myDUMP > /dev/null

# Restore index patterns
myINDEXID=$(ls patterns/*.json | cut -c 10- | rev | cut -c 6- | rev)
myINDEXCOUNT=$(cat patterns/$myINDEXID.json | tr '\\' '\n' | grep "scripted" | wc -w)
echo $myCOL1"### Now importing"$myCOL0 $myINDEXCOUNT $myCOL1"index pattern fields." $myCOL0
$myCURL -XDELETE ''$myKIBANA'api/saved_objects/index-pattern/logstash-*' -H "Content-Type: application/json" -H "kbn-xsrf: true" > /dev/null
$myCURL -XDELETE ''$myKIBANA'api/saved_objects/index-pattern/'$myINDEXID'' -H "Content-Type: application/json" -H "kbn-xsrf: true" > /dev/null
$myCURL -XPOST ''$myKIBANA'api/saved_objects/index-pattern/'$myINDEXID'' -H "Content-Type: application/json" -H "kbn-xsrf: true" -d @patterns/$myINDEXID.json > /dev/null &
echo

# Restore dashboards
myDASHBOARDS=$(ls dashboards/*.json | cut -c 12- | rev | cut -c 6- | rev) 
echo $myCOL1"### Now importing "$myCOL0$(echo $myDASHBOARDS | wc -w)$myCOL1 "dashboards." $myCOL0
for i in $myDASHBOARDS;
  do
    $myCURL -XDELETE ''$myKIBANA'api/saved_objects/dashboard/'$i'' -H "Content-Type: application/json" -H "kbn-xsrf: true" > /dev/null &
  done;
wait
for i in $myDASHBOARDS;
  do
    echo $myCOL1"###### "$i $myCOL0
    $myCURL -XPOST ''$myKIBANA'api/saved_objects/dashboard/'$i'' -H "Content-Type: application/json" -H "kbn-xsrf: true" -d @dashboards/$i.json > /dev/null &
  done;
wait
echo

# Restore visualizations
myVISUALIZATIONS=$(ls visualizations/*.json | cut -c 16- | rev | cut -c 6- | rev)
echo $myCOL1"### Now importing "$myCOL0$(echo $myVISUALIZATIONS | wc -w)$myCOL1 "visualizations." $myCOL0
for i in $myVISUALIZATIONS;
  do
    $myCURL -XDELETE ''$myKIBANA'api/saved_objects/visualization/'$i'' -H "Content-Type: application/json" -H "kbn-xsrf: true" > /dev/null &
  done;
wait
for i in $myVISUALIZATIONS;
  do
    echo $myCOL1"###### "$i $myCOL0
    $myCURL -XPOST ''$myKIBANA'api/saved_objects/visualization/'$i'' -H "Content-Type: application/json" -H "kbn-xsrf: true" -d @visualizations/$i.json > /dev/null &
  done;
wait
echo

# Restore searches
mySEARCHES=$(ls searches/*.json | cut -c 10- | rev | cut -c 6- | rev) 
echo $myCOL1"### Now importing "$myCOL0$(echo $mySEARCHES | wc -w)$myCOL1 "searches." $myCOL0
for i in $mySEARCHES;
  do
    $myCURL -XDELETE ''$myKIBANA'api/saved_objects/search/'$i'' -H "Content-Type: application/json" -H "kbn-xsrf: true" > /dev/null &
  done;
wait
for i in $mySEARCHES;
  do
    echo $myCOL1"###### "$i $myCOL0
    $myCURL -XPOST ''$myKIBANA'api/saved_objects/search/'$i'' -H "Content-Type: application/json" -H "kbn-xsrf: true" -d @searches/$i.json > /dev/null &
  done;
echo
wait

# Restore configs
myCONFIGS=$(ls configs/*.json | cut -c 9- | rev | cut -c 6- | rev)
echo $myCOL1"### Now importing "$myCOL0$(echo $myCONFIGS | wc -w)$myCOL1 "configs." $myCOL0
for i in $myCONFIGS;
  do
    $myCURL -XDELETE ''$myKIBANA'api/saved_objects/config/'$i'' -H "Content-Type: application/json" -H "kbn-xsrf: true" > /dev/null &
  done;
wait
for i in $myCONFIGS;
  do
    echo $myCOL1"###### "$i $myCOL0
    $myCURL -XPOST ''$myKIBANA'api/saved_objects/config/'$i'' -H "Content-Type: application/json" -H "kbn-xsrf: true" -d @configs/$i.json > /dev/null &
  done;
echo
wait

# Stats
echo
echo $myCOL1"### Statistics"
echo $myCOL1"###### Imported"$myCOL0 $myINDEXCOUNT $myCOL1"index patterns." $myCOL0
echo $myCOL1"###### Imported"$myCOL0 $(echo $myDASHBOARDS | wc -w) $myCOL1"dashboards." $myCOL0
echo $myCOL1"###### Imported"$myCOL0 $(echo $myVISUALIZATIONS | wc -w) $myCOL1"visualizations." $myCOL0
echo $myCOL1"###### Imported"$myCOL0 $(echo $mySEARCHES | wc -w) $myCOL1"searches." $myCOL0
echo $myCOL1"###### Imported"$myCOL0 $(echo $myCONFIGS | wc -w) $myCOL1"configs." $myCOL0
echo



sw-runner.sh
#!/bin/bash

# Got root?
myWHOAMI=$(whoami)
if [ "$myWHOAMI" != "root" ]
  then
    echo "### Need to run as root ..."
    echo
    exit
fi

### Define Slack-Watchman runner
fuSWRUN () {
local SLACK_WATCHMAN_NAME="$1"
local SLACK_WATCHMAN_TOKEN="$2"
local SLACK_WATCHMAN_LOGFILE="$(echo $SLACK_WATCHMAN_NAME | tr -d " ").json"

# [--users] [--channels] [--pii] [--financial] [--tokens] [--files] [--custom CUSTOM]
local SLACK_WATCHMAN_QUERY="--all"
# --timeframe {d,w,m,a}
local SLACK_WATCHMAN_TIMEFRAME="--timeframe a"

echo -n "### Now running: $SLACK_WATCHMAN_NAME"
docker run --name slack-watchman \
           --rm \
           --read-only \
            -v /data/slack-watchman/log:/opt/slack-watchman/log:rw \
           --env SLACK_WATCHMAN_TOKEN="$SLACK_WATCHMAN_TOKEN" \
           --env SLACK_WATCHMAN_LOGFILE="$SLACK_WATCHMAN_LOGFILE" \
           --env SLACK_WATCHMAN_QUERY="$SLACK_WATCHMAN_QUERY" \
           --env SLACK_WATCHMAN_TIMEFRAME="$SLACK_WATCHMAN_TIMEFRAME" \
           dtagdevsec/slack-watchman:latest
echo " - Done."
echo
}

# fuSWRUN "Workspace_Name" "OAuthToken"
fuSWRUN "Workspace_Name" "xoxp-1111111111111-2222222222222-3333333333333-44444444444444444444444444444444"
